File: app/src/main/java/io/sweers/catchup/injection/ConductorInjection.java
package io.sweers.catchup.injection;

import android.app.Activity;
import android.util.Log;
import com.bluelinelabs.conductor.Controller;
import dagger.android.AndroidInjector;
import dagger.android.DispatchingAndroidInjector;

import static dagger.internal.Preconditions.checkNotNull;

/** Injects core Conductor types. */
public class ConductorInjection {
  private static final String TAG = "io.sweers.catchup.ConductorInjection";

  /**
   * Injects {@code controller} if an associated {@link AndroidInjector.Factory} implementation can be
   * found, otherwise throws an {@link IllegalArgumentException}.
   *
   * <p>Uses the following algorithm to find the appropriate {@code
   * DispatchingAndroidInjector<Controller>} to inject {@code controller}:
   *
   * <ol>
   * <li>Walks the parent-controller hierarchy to find the a controller that implements {@link
   * HasDispatchingControllerInjector}, and if none do
   * <li>Uses the {@code controller}'s {@link Controller#getActivity() activity} if it implements
   * {@link HasDispatchingControllerInjector}, and if not
   * <li>Uses the {@link android.app.Application} if it implements {@link
   * HasDispatchingControllerInjector}.
   * </ol>
   *
   * If none of them implement {@link HasDispatchingControllerInjector}, a {@link
   * IllegalArgumentException} is thrown.
   *
   * @throws IllegalArgumentException if no {@code AndroidInjector.Factory<Controller, ?>} is bound
   * for {@code controller}.
   */
  public static void inject(Controller controller) {
    checkNotNull(controller, "controller");
    HasDispatchingControllerInjector hasDispatchingControllerInjector =
        findHasControllerInjector(controller);
    Log.d(TAG, String.format(
        "An injector for %s was found in %s",
        controller.getClass()
            .getCanonicalName(),
        hasDispatchingControllerInjector.getClass()
            .getCanonicalName()));

    DispatchingAndroidInjector<Controller> controllerInjector =
        hasDispatchingControllerInjector.controllerInjector();
    checkNotNull(
        controllerInjector,
        "%s.controllerInjector() returned null",
        hasDispatchingControllerInjector.getClass()
            .getCanonicalName());

    controllerInjector.inject(controller);
  }

  private static HasDispatchingControllerInjector findHasControllerInjector(Controller controller) {
    Controller parentController = controller;
    while ((parentController = parentController.getParentController()) != null) {
      if (parentController instanceof HasDispatchingControllerInjector) {
        return (HasDispatchingControllerInjector) parentController;
      }
    }
    Activity activity = controller.getActivity();
    if (activity instanceof HasDispatchingControllerInjector) {
      return (HasDispatchingControllerInjector) activity;
    }
    if (activity.getApplication() instanceof HasDispatchingControllerInjector) {
      return (HasDispatchingControllerInjector) activity.getApplication();
    }
    throw new IllegalArgumentException(String.format(
        "No injector was found for %s",
        controller.getClass()
            .getCanonicalName()));
  }
}

File: app/src/main/kotlin/io/sweers/catchup/injection/ConductorInjection.kt
/*
 * Copyright (c) 2018 Zac Sweers
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.sweers.catchup.injection

import com.bluelinelabs.conductor.Controller
import dagger.android.AndroidInjector
import dagger.android.HasActivityInjector
import dagger.internal.Preconditions.checkNotNull
import io.sweers.catchup.util.d

/** Injects core Conductor types.  */
object ConductorInjection {

  /**
   * Injects `controller` if an associated [AndroidInjector.Factory] implementation can be
   * found, otherwise throws an [IllegalArgumentException].
   *
   * Uses the following algorithm to find the appropriate `DispatchingAndroidInjector<Controller>` to inject `controller`:
   *
   *  1. Walks the parent-controller hierarchy to find the a controller that implements [HasControllerInjector], and if none do
   *  2. Uses the `controller`'s [activity][Controller.getActivity] if it implements
   * [HasControllerInjector], and if not
   *  3. Uses the [android.app.Application] if it implements [HasActivityInjector].
   *
   * If none of them implement [HasControllerInjector], a [IllegalArgumentException] is thrown.
   *
   * @throws IllegalArgumentException if no `AndroidInjector.Factory<Controller, ?>` is bound
   * for `controller`.
   */
  fun inject(controller: Controller) {
    val hasDispatchingControllerInjector = findHasControllerInjector(controller)
    d {
      String.format(
          "An injector for %s was found in %s",
          controller.javaClass
              .canonicalName,
          hasDispatchingControllerInjector.javaClass
              .canonicalName)
    }

    val controllerInjector = hasDispatchingControllerInjector.controllerInjector()
    checkNotNull(
        controllerInjector,
        "%s.controllerInjector() returned null",
        hasDispatchingControllerInjector.javaClass
            .canonicalName)

    controllerInjector.inject(controller)
  }

  private fun findHasControllerInjector(controller: Controller): HasControllerInjector {
    val controllerSequence = generateSequence(controller) { it.parentController } // <3 Kotlin
    controllerSequence.forEach {
      if (it is HasControllerInjector) {
        return it
      }
    }
    controller.activity?.let {
      if (it is HasControllerInjector) {
        return it
      } else if (it.application is HasControllerInjector) {
        return it.application as HasControllerInjector
      }
    }
    throw IllegalArgumentException(String.format(
        "No injector was found for %s",
        controller.javaClass
            .canonicalName))
  }
}
